drv/transceivers-server/src/udp.rs

```diff
@@ -217,397 +215,572 @@ impl ServerImpl {
-                ) {
-                    Ok(r) => r,
-                    Err(e) => (SpResponse::Error(e), 0),
-                }
+            MessageBody::HostRequest(h) => self.handle_host_request(
+                h,
+                data,
+                &mut tx_data_buf[reserved_framing..],
+            ),
+            // Nothing implemented yet
+            MessageBody::Error(e) => {
+                ringbuf_entry!(Trace::GotError(e));
+                return None;
             }
         };
 
-        // Serialize the Message into the front of the tx buffer.
-        // Note that we send out _our_ protocol version number, which may differ
-        // from the host's, but should be compatible with it.
-        let header = Header {
-            version: version::CURRENT,
-            ..msg.header
-        };
-        let out = Message {
-            header,
-            modules: msg.modules,
-            body: MessageBody::SpResponse(reply),
-        };
-        let msg_len = hubpack::serialize(tx_data_buf, &out).unwrap();
+        // Serialize the Header into the front of the tx buffer, followed by
+        // the actual Message. This sends out _our_ protocol version number,
+        // which may differ from the host's, but should be compatible with it.
+        let response = Message::new(body);
+        let header = Header::new(msg_id, response.kind());
+
+        let hdr_len = hubpack::serialize(tx_data_buf, &header).unwrap();
+        let msg_len =
+            hubpack::serialize(&mut tx_data_buf[hdr_len..], &response).unwrap();
+
+        ringbuf_entry!(Trace::ResponseSize(ResponseSize {
+            header_length: hdr_len as u8,
+            message_length: msg_len as u8,
+            data_length: data_len as u16
+        }));
 
         // At this point, any supplementary data was written to
         // `tx_data_buf[Message::MAX_SIZE..]`, so it's not necessarily tightly
         // packed against the end of the `Message`.  Let's shift it backwards
         // based on the size of the leading `Message`:
         tx_data_buf.copy_within(
-            Message::MAX_SIZE..(Message::MAX_SIZE + data_len),
-            msg_len,
+            reserved_framing..(reserved_framing + data_len),
+            hdr_len + msg_len,
         );
-        Some((msg_len + data_len) as u32)
+        Some((hdr_len + msg_len + data_len) as u32)
     }
 
     fn handle_host_request(
         &mut self,
         h: HostRequest,
-        modules: ModuleId,
         data: &[u8],
         out: &mut [u8],
-    ) -> Result<(SpResponse, usize), Error> {
+    ) -> (MessageBody, usize) {
         match h {
-            HostRequest::Status => {
+            HostRequest::Status(modules) => {
                 ringbuf_entry!(Trace::Status(modules));
-                let count = self.get_status(modules, out)?;
-                Ok((SpResponse::Status, count))
+                let mask = LogicalPortMask(modules.0 as u32);
```

Comment:
Might be useful to `impl From<ModuleId> for LogicalPortMask`, to avoid the casting in many places.